### ISSUE TITLE
Implement tag system

### ISSUE THREAD

**Issue by martinruefenacht** [mpi-forum/pympistandard#7]
We want to implement a tag system with which we can tag procedures. Then we can add functionality which the emitter would use to emit lists for example.

The tags should be provided access to via a property (tags).

@GuillaumeMercier I remember you needed this. Do you have any comments?

**Comment by GuillaumeMercier** [comment]
No I don't, because I'm unsure on how I would be using this feature. But I need this (and probably others too).

**Comment by martinruefenacht** [comment]
The intention is to have an additional user function that can be used in the mpi-binding blocks:
```python
    function_name("MPI_Send")
    tag("blocking")
    
    parameter(
        "buf", "BUFFER", desc="initial address of send buffer", constant=True
    )
    parameter(
        "count",
        "POLYXFER_NUM_ELEM_NNI",
        desc="number of elements in send buffer",
    )
    parameter(
        "datatype", "DATATYPE", desc="datatype of each send buffer element"
    )
    parameter("dest", "RANK", desc="rank of destination")
    parameter("tag", "TAG", desc="message tag")
    parameter("comm", "COMMUNICATOR")
```

And then within the latex you have another "macro" `\listtagged{"blocking"}`, this would emit a list of "MPI_Send, MPI_Recv, MPI_Test, and MPI_Wait" for example. This list of latex macros can be expanded of course.

In fact it would be great to have a list of macros people want to have.

**Comment by GuillaumeMercier** [comment]
> In fact it would be great to have a list of macros people want to have.

A list of tags you mean?

**Comment by GuillaumeMercier** [comment]
Or maybe your `\listtagged` macros could accept several tags? E.g. `\listtagged{"blocking","communicator_creation"}` ?

**Comment by martinruefenacht** [comment]
> > In fact it would be great to have a list of macros people want to have.
> 
> A list of tags you mean?

No, the tags in the current idea is just strings, and no standardization. That allows authors to do anything without having to centralize. Yes, we can make them accept several tags with two different modes, either union (print every element of those sets) or intersection (print only those elements that are in all sets).

I imagine most times it will be chapter authors having some "meaningful" name for a group of procedures that they later want to reference with a listing somewhere. This was at least the initial idea, but of course tags could be much more powerful.

**Comment by GuillaumeMercier** [comment]
I missed the `tag("blocking")` line in your example. I see what you mean.
